[Bolt] Fix address translation for KASLR kernel

xur-llvm opened the pull request:
This patch enables Bolt to analyze kernel addresses that have been randomized by KASLR. It parses memory map (MMap) entries within perf files to find the address mapping. Note that that kernel modules are still not handled.

xur-llvm committed:
[Bolt] Fix address translation for KASLR kernel

This patch enables Bolt to analyze kernel addresses that
have been randomized by KASLR. It parses memory map (MMap)
entries within perf files to find the address mapping.

maksfb commented:
I haven't checked yet, but this PR should fix #99373.

maksfb commented:
Thank you for the patch, Ron (@xur-llvm). I verified that it works for `vmlinux` with KASLR. Sadly, at the same time it breaks non-KASLR `vmlinux` processing for `perf2bolt`.

xur-llvm commented:
What is the symptom of the breakage? Can you share the perf file and the
vmlinux so that I can debug?

I just tried my patch on 6.12.0-rc6-1 for non-KASLR profile and it
generated exactly the same bolt profile using perf2bolt without my patch.
This is on arch-linux.

-Rong

On Sun, Nov 10, 2024 at 7:34 PM Maksim Panchenko ***@***.***>
wrote:

> Thank you for the patch, Ron ***@***.*** <https://github.com/xur-llvm>). I
> verified that it works for vmlinux with KASLR. Sadly, at the same time it
> breaks non-KASLR vmlinux processing for perf2bolt.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/llvm/llvm-project/pull/114261#issuecomment-2467164591>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AOI42XTLY4HDHJ25LEZ3WAD2AAQUPAVCNFSM6AAAAABQ4Q6YGGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDINRXGE3DINJZGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

maksfb commented:
> What is the symptom of the breakage? Can you share the perf file and the vmlinux so that I can debug? I just tried my patch on 6.12.0-rc6-1 for non-KASLR profile and it generated exactly the same bolt profile using perf2bolt without my patch. 

On your system you must have matched `vmlinux` binary by build-id. When build-id doesn't match or doesn't exist, with this patch `perf2bolt` will try to match the kernel binary by name. In `perf.data` it's recorded as "[kernel.kallsyms]" hence the error from `perf2bolt`. The workaround is to rename "vmlinux" to "[kernel.kallsyms]" which I'd rather avoid since unlike in the user space there's no ambiguity about the kernel binary.

maksfb commented on bolt/lib/Rewrite/RewriteInstance.cpp:
```diff
@@ -530,8 +530,11 @@ Error RewriteInstance::discoverStorage() {
-          Phdr.p_vaddr >= BinaryContext::KernelStartX86_64)
+          Phdr.p_vaddr >= BinaryContext::KernelStartX86_64) {
         BC->IsLinuxKernel = true;
+        BC->HasFixedLoadAddress = false;
```
Is there any way to detect KASLR by looking at `vmlinux` ELF and only then set `HasFixedLoadAddress`.

xur-llvm commented:
OK. I see. This should be easy to fix: the logic of detecting kernel is not
using the name. So if this is kernel, we can skip the match and use the
provided ELF file directly.
Or, we can pass a flat for kernel? (I don't think you would like this).

-Rong

On Mon, Nov 11, 2024 at 9:16 PM Maksim Panchenko ***@***.***>
wrote:

> What is the symptom of the breakage? Can you share the perf file and the
> vmlinux so that I can debug? I just tried my patch on 6.12.0-rc6-1 for
> non-KASLR profile and it generated exactly the same bolt profile using
> perf2bolt without my patch.
>
> On your system you must have matched vmlinux binary by build-id. When
> build-id doesn't match or doesn't exist, with this patch perf2bolt will
> try to match the kernel binary by name. In perf.data it's recorded as
> "[kernel.kallsyms]" hence the error from perf2bolt. The workaround is to
> rename "vmlinux" to "[kernel.kallsyms]" which I'd rather avoid since unlike
> in the user space there's no ambiguity about the kernel binary.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/llvm/llvm-project/pull/114261#issuecomment-2469628863>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AOI42XS7H4FGILEIM76442L2AGFLJAVCNFSM6AAAAABQ4Q6YGGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDINRZGYZDQOBWGM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

xur-llvm commented:
When I worked on this patch, I had a question on why
HasFixedLoadAddress is needed. To me, ASLR (KASLR), or not, should use
one code path. Non-ASLR (to me) is a trivial case of ASLR -- at least
from parsing and reading perf file point of view.
I was thinking about removing this field. But this field also used to
control some optimizations. So I leave it there.

It seems to me the only upside of using FixedLoadAddress is to avoid
address adjustments -- but that should be very cheap.

Is my understanding correct?

Thanks,

-Rong

On Mon, Nov 11, 2024 at 9:20 PM Maksim Panchenko
***@***.***> wrote:
>
> @maksfb commented on this pull request.
>
> ________________________________
>
> In bolt/lib/Rewrite/RewriteInstance.cpp:
>
> >          BC->IsLinuxKernel = true;
> +        BC->HasFixedLoadAddress = false;
>
> Is there any way to detect KASLR by looking at vmlinux ELF and only then set HasFixedLoadAddress.
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.Message ID: ***@***.***>

maksfb commented:
> When I worked on this patch, I had a question on why HasFixedLoadAddress is needed. To me, ASLR (KASLR), or not, should use one code path. Non-ASLR (to me) is a trivial case of ASLR -- at least from parsing and reading perf file point of view. I was thinking about removing this field. But this field also used to control some optimizations. So I leave it there. It seems to me the only upside of using FixedLoadAddress is to avoid address adjustments -- but that should be very cheap. Is my understanding correct? Thanks

One example where we use `HasFixedLoadAddress` flag is in the user-space to generate slightly different EH tables for PIEs (i.e. when flag is set to `false`) as we cannot use absolute addressing. I imagine it could be similar for `vmlinux` where KASLR-capable kernel requires all addressing to be position-independent and thus some metadata cannot be absolute addresses. Most of the metadata BOLT detects uses relative addressing with the exception of one field in `.parainstructions`. The latter is no longer used on modern kernels, but it will be good to know if an older kernel is KASLR-capable and reject processing `.parainstructions` for it.

maksfb commented:
> OK. I see. This should be easy to fix: the logic of detecting kernel is not using the name. So if this is kernel, we can skip the match and use the provided ELF file directly. Or, we can pass a flat for kernel? (I don't think you would like this).

Sound like a plan. No flag is needed. You currently check for "[kernel.kallsyms]" in the `perf script` output and it seems to work on the systems I've checked. But I wonder it's 100% reliable? The kernel also appears on the first line and the only binary with non-zero "@" field in `PERF_RECORD_MMAP`.

xur-llvm committed:
Merge branch 'main' into users/xur-llvm/spr/bolt

xur-llvm committed:
[bolt] Use a fixed name for the kernel image

Use a fixed name for the kernel image to process kernel profiles,
regardless of the presence of a build ID or not.

This is to address the issue when the provided kernel image lacks a
matching build ID. This name, "[kernel.kallsyms]", is the default
for kernel DSOs in the Linux kernel source code
(see https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/tools/perf/util/dso.c#n428).

While "[guest.kernel.kallsyms]" is the kernel DSO name for guest kernel,
support for VM profiles is currently limited. Therefore, we can
skip this name for now.

xur-llvm commented:
@maksfb: Maksim, could you tell the updated patch to see if it works for the kernel image without build-id?

maksfb commented:
> @maksfb: Maksim, could you tell the updated patch to see if it works for the kernel image without build-id?

Thanks. I'm back from vacation. Will test it.
